File: lib/dotfiles/command_runner.rb
# frozen_string_literal: true

require 'open3'

module Dotfiles
    # Runs commands while preserving interactive terminal input when needed.
    class CommandRunner
        # Signals that an external command could not complete successfully.
        class Failure < StandardError
        end

        # Runs a command while capturing its output for inspection.
        #
        # @param command [Array<String>] executable and arguments
        # @return [String] standard output from the command
        # @raise [Failure] when the command exits unsuccessfully
        def capture(command)
            stdout, stderr, status = Open3.capture3(*command)
            unless status.success?
                output = stderr.empty? ? stdout : stderr
                message = if output.strip.empty?
                              "command failed (#{status.exitstatus || 'unknown'}): #{command.join(' ')}"
                          else
                              output
                          end
                raise Failure, message
            end

            stdout
        end

        # Runs a command with the current terminal attached for interaction.
        #
        # @param command [Array<String>] executable and arguments
        # @return [void]
        # @raise [Failure] when the command exits unsuccessfully
        def interactive(command)
            raise Failure, "command failed: #{command.join(' ')}" unless system(*command)
        end
    end
end

File: src/dotfiles/command_runner.py
"""External command execution with captured and interactive modes."""

from __future__ import annotations

import subprocess
from collections.abc import Mapping, Sequence
from pathlib import Path


class CommandFailure(RuntimeError):
    """Indicate that an external command returned a failure status."""


class CommandRunner:
    """Run commands without invoking a shell."""

    def capture(
        self,
        command: Sequence[str],
        cwd: Path | None = None,
        env: Mapping[str, str] | None = None,
    ) -> str:
        """Return standard output or raise with the command failure details."""
        completed = subprocess.run(
            list(command),
            cwd=cwd,
            env=env,
            capture_output=True,
            text=True,
            check=False,
        )
        if completed.returncode == 0:
            return completed.stdout

        output = completed.stderr or completed.stdout
        message = output if output.strip() else f"command failed ({completed.returncode}): {' '.join(command)}"

        raise CommandFailure(message)

    def interactive(
        self,
        command: Sequence[str],
        cwd: Path | None = None,
        env: Mapping[str, str] | None = None,
    ) -> None:
        """Run a command with the current terminal attached."""
        completed = subprocess.run(list(command), cwd=cwd, env=env, check=False)

        if completed.returncode != 0:
            raise CommandFailure(f"command failed ({completed.returncode}): {' '.join(command)}")
